### PR TITLE
KAFKA-14538: Implement KRaft metadata transactions

### PR DESCRIPTION
**Pull request by cmccabe**
Implement KRaft metadata transactions as specified in KIP-868.

This includes creating a new metadata version, IBP_3_6_IV1, which contains the three new records: AbortTransactionRecord, BeginTransactionRecord, EndTransactionRecord.

Maintaining the various offsets and in-memory snapshots in QuorumController was getting too complex. Therefore, this PR splits that logic out into the LogReplayTracker class. This makes it possible to unit test events like Raft leader changes, which is difficult to do when the code is directly in QuorumController.java.

### CODE REVIEW

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java [hunk 1]
```diff
@@ -267,6 +270,10 @@ public boolean isLeaderEpochBumpRequiredOnIsrShrink() {
         return !this.isAtLeast(IBP_3_6_IV0);
     }
 
+    public boolean isMetadataTransactionSupported() {
+        return !this.isAtLeast(IBP_3_6_IV1);
```
Comment (mumrah) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:274]: This boolean is inverted

File: metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java [hunk 2]
```diff
@@ -17,56 +17,516 @@
...
-    public static class Builder {
+class LogReplayTracker {
+    static class Builder {
         private LogContext logContext = null;
+        private Time time = Time.SYSTEM;
+        private int nodeId = -1;
+        private QuorumControllerMetrics metrics = null;
+        private Consumer<Integer> gainLeadershipCallback = __ -> { };
+        private Consumer<Integer> loseLeadershipCallback = __ -> { };
+        private BiConsumer<Long, String> revertCallback = (__, ___) -> { };
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java:49]: WDYT about pulling all these callbacks into a single interface?

File: metadata/src/main/java/org/apache/kafka/controller/QuorumController.java [hunk 3]
```diff
@@ -1915,19 +1809,24 @@ private QuorumController(
...
-        this.logReplayTracker = new LogReplayTracker.Builder().
+        this.tracker = new LogReplayTracker.Builder().
             setLogContext(logContext).
+            setTime(time).
+            setNodeId(nodeId).
+            setMetrics(controllerMetrics).
+            setGainLeadershipCallback(this::gainLeadership).
+            setLoseLeadershipCallback(this::loseLeadership).
+            setFatalFaultHandler(fatalFaultHandler).
+            setRevertCallback(this::revert).
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:1820]: Missing the snapshot registry here

File: metadata/src/main/java/org/apache/kafka/controller/QuorumController.java [hunk 4]
```diff
@@ -1309,48 +1221,57 @@ public void processBatchEndOffset(long offset) {
...
+     *
+     * @param offset    The offset of the in-memory snapshot.
+     * @param purpose   The purpose of the loading process.
+     */
+    private void finishLoadingInMemorySnapshot(long offset, String purpose) {
+        if (tracker.active()) {
+            snapshotRegistry.getOrCreateSnapshot(offset);
         }
+        authorizer.ifPresent(a -> a.loadSnapshot(aclControlManager.idToAcl()));
+        log.info("FInished loading in-memory snapshot at {} after {]", offset, purpose);
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:1274]:
```suggestion
        log.info("Finished loading in-memory snapshot at {} after {}", offset, purpose);
```

File: metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java [hunk 5]
```diff
@@ -17,56 +17,516 @@
...
+
+    void replay(BeginTransactionRecord record, long offset) {
+        if (transactionStartOffset != -1L) {
+            throw fatalFaultHandler.handleFault("tried to begin metadata transaction at " + offset +
+                ", but we are still inside a metadata transaction starting at " +
+                transactionStartOffset);
+        }
+        log.debug("Beginning metadata transaction at offset {}{}.", offset,
+            record.name() == null ? "" : " named " + record.name());
+        snapshotRegistry.getOrCreateSnapshot(offset);
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java:492]: Is it possible to move all the SnapshotRegistry mutating calls (getOrCreateSnapshot, deleteSnapshots, revertToSnapshot) into LogReplayTracker? It's a little hard to follow the in-memory snapshotting when it's split between here and QuorumController?

**Review by mumrah — COMMENTED** [review]
Thanks @cmccabe! Since this PR is already pretty big without the MetadataLoader changes, what do you think about merging this as-is (after reviews) and add the remaining changes as follow-on PRs? Otherwise, I worry it will grow really large and become hard to review.

**Comment by cmccabe** [comment]
> Thanks @cmccabe! Since this PR is already pretty big without the MetadataLoader changes, what do you think about merging this as-is (after reviews) and add the remaining changes as follow-on PRs? Otherwise, I worry it will grow really large and become hard to review.

Fair point. Yeah, let's do that.

File: metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java (hunk 2 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java:49]: ok

**Comment by cmccabe** [comment]
duplicate of #14200
